doc/sfputil/read_write_eeprom_by_page.md

```diff
@@ -0,0 +1,250 @@
+- Support reading/writing cable EEPROM for all types of cables except RJ45.
+- Support non flat memory mode reading/writing. In this mode, user shall proved page and offset according to the standard. For example, CMIS page 1h starting offset is 128, offset less than 128 shall be treated as invalid.
+- Support flat memory mode reading/writing (Phase 2). In this mode, the EEPROM memory shall be treated as continues flat memory. This design document will not discuss this mode.
+- Vendor who does not support `sfp.read_eeprom` and `sfp.write_eeprom` is expected to raise `NotImplementedError`, this error shall be properly handled
+- Others error shall be treated as read/write failure
+- Vendor specific implementation is not required but still possible.
+
+
+### Architecture Design
+
+The current architecture is not changed.
+
+### High-Level Design
+
+Submodule sonic-platform-common and sonic-utilities shall be extended to support this feature. Vendor specific implementation is not required but still possible.
+
+#### sonic-platform-common
+
+The existing API `sfp.read_eeprom` and `sfp.write_eeprom` accept "overall offset" as parameter. They have no concept of page. If user wants to use it, user has to convert page and offset to "overall offset" manually. Different cable types use different covert method according to the standard. So, it is not user friendly to provide such API directly to user. Two new API shall be added to support this feature:
+
+```python
+
+def read_eeprom_by_page(self, page, offset, size, wire_addr=None, flat=False):
+    """
+    Read EEPROM by page
+
+    Args:
+        page: EEPROM page number. Raise ValueError for invalid page.
+        offset: EEPROM page offset. Raise ValueError for invalid offset.
+        size: Read size. Raise ValueError for invalid size.
+        wire_addr: Wire address. Only valid for sff8472. Raise ValueError for invalid wire address.
+        flat: Read mode.
+
+    Returns:
+        A string contains the hex format EEPROM data.
+    """
+    raise NotImplementedError
+
+def write_eeprom_by_page(self, page, offset, data, wire_addr=None, flat=False):
+    """
+    Write EEPROM by page
+
+    Args:
+        page: EEPROM page number. Raise ValueError for invalid page.
+        offset: EEPROM page offset. Raise ValueError for invalid offset.
+        data: Binary EEPROM data.
+        wire_addr: Wire address. Only valid for sff8472. Raise ValueError for invalid wire address.
+        flat: Write mode.
+
+    Returns:
+        True if write successfully else False
+    """
+    raise NotImplementedError
+```
+
+`sfp_optoe_base.SfpOptoeBase` shall implement these new APIs so that vendor does not have to implement it (Vendor specific implementation is also possible).
+
+```python
+def read_eeprom_by_page(self, page, offset, size, wire_addr=None, flat=False):
+    api = self.get_xcvr_api()
+    overall_offset = api.get_overall_offset(page, offset, size, wire_addr, flat) if api is not None else None
```

Comment:
@Junchao-Mellanox do we have PR for get_overall_offset()?

Reply:
Yes, it is under test, I will create PR once it is ready.